### PR TITLE
Implement UI errata tests

### LINKED ISSUE #3673: Implement UI errata stubs

### PR DESCRIPTION
**Pull request by abalakh**
Closes #3673

### CODE REVIEW

**Comment by abalakh** [comment]
`[WIP]` as i don't have test results yet (and as writing these tests took more than a week i'm pretty sure something became broken, i'm already aware of at least one test failing), but overall PR is ready for review.

**Comment by oshtaier** [comment]
Remove that '4,538' long txt file please =)

File: robottelo/constants.py [hunk 1]
```diff
@@ -241,7 +241,7 @@
 DEFAULT_CV = "Default Organization View"
 DEFAULT_PTABLE = "Kickstart default"
 DEFAULT_SUBSCRIPTION_NAME = (
-    'Red Hat Enterprise Linux Server Entry Level, Self-support')
+    'Red Hat Enterprise Linux Server, Premium (Physical or Virtual Nodes)')
```
Comment (oshtaier) [inline robottelo/constants.py]: That a little bit risky change that will require running some more tests probably

File: robottelo/ui/locators/common.py [hunk 2]
```diff
@@ -75,7 +75,9 @@
 
     "search": (By.ID, "search"),
     "auto_search": (
-        By.XPATH, "//ul[contains(@id, 'ui-id')]/li/a[contains(., '%s')]"),
+        By.XPATH,
```
Comment (oshtaier) [inline robottelo/ui/locators/common.py]: That will also require running auto-complete tests for organization and location

**Comment by oshtaier** [comment]
And runs from `tests/foreman/ui/test_hostcollection.py` too

**Comment by oshtaier** [comment]
Basically ACK requiring a lot of test runs

File: robottelo/constants.py (hunk 1 above)
Comment (abalakh) [inline robottelo/constants.py]: Yeah. Well as i already have to rerun hostcollection tests i think that's the way we'll validate the change. Anyway it works perfectly with existing contenthost registering/errata installation tests, and right now all the subscriptions-related tests are failing on jenkins, so this subscription should be a good candidate.

File: robottelo/ui/locators/common.py (hunk 2 above)
Comment (abalakh) [inline robottelo/ui/locators/common.py]: I've checked manually if the new xpath is ok with organizations but haven't taken a look on location, will do that. Long story short previous xpath was working only with foreman-style search bar, new one should work with both foreman and katello styles.

**Comment by abalakh** [comment]
Test results:
First attempt:
```python
============================= test session starts ==============================
platform linux2 -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1, repeat-0.2
collecting ... collected 17 items

tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_apply_for_all_hosts FAILED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_apply_for_host PASSED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_chost_library PASSED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_chost_previous_env PASSED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_filter FAILED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_incremental_update <- ../../env/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_list FAILED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_list_permission PASSED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_search_autocomplete PASSED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_search_redirect <- ../../env/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_show_count_on_chost_details_page FAILED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_show_count_on_chost_page PASSED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_sort <- ../../env/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_view <- ../../env/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_view_cve PASSED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_view_details <- ../../env/lib/python2.7/site-packages/unittest2/case.py SKIPPED
tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_view_products_and_repos PASSED
=============== 4 failed, 8 passed, 5 skipped in 3421.31 seconds ===============
```
Fixed failing tests and reran them:
```python
============================= test session starts ==============================
platform linux2 -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1, repeat-0.2
collecting ... collected 17 items

tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_apply_for_all_hosts PASSED

========= 16 tests deselected by '-ktest_positive_apply_for_all_hosts' =========
================== 1 passed, 16 deselected in 1125.30 seconds ==================
============================= test session starts ==============================
platform linux2 -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1, repeat-0.2
collecting ... collected 17 items

tests/foreman/ui/test_errata.py::ErrataTestCase::test_positive_show_count_on_chost_details_page PASSED

== 16 tests deselected by '-ktest_positive_show_count_on_chost_details_page' ===
================== 1 passed, 16 deselected in 1423.54 seconds ==================
```
Note that because of issue with subscriptions for custom products not appearing (i think it's tracked in #3642 and maybe [BZ1360239](https://bugzilla.redhat.com/show_bug.cgi?id=1360239) is related too) the tests are randomly failing as well as all the rest our t3 tests on CI (e.g., host package/errata management, remote execution, host collection pkg management tests etc). It's pretty important issue with either our automation or application, but it's not related to this pull request specifically.

**Comment by abalakh** [comment]
Test results for existing search autocomplete tests:
```python
================================================ test session starts ================================================
platform linux2 -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1, repeat-0.2
collected 33 items 

tests/foreman/ui/test_organization.py::OrganizationTestCase::test_positive_search_autocomplete PASSED

=========================== 32 tests deselected by '-ktest_positive_search_autocomplete' ============================
===================================== 1 passed, 32 deselected in 77.25 seconds ======================================
py.test tests/foreman/ui/test_location.py -v -k 'test_positive_auto_search'
================================================ test session starts ================================================
platform linux2 -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1, repeat-0.2
collected 27 items 

tests/foreman/ui/test_location.py::LocationTestCase::test_positive_auto_search <- robottelo/decorators/__init__.py PASSED

=============================== 26 tests deselected by '-ktest_positive_auto_search' ================================
===================================== 1 passed, 26 deselected in 47.70 seconds ======================================
```

File: tests/foreman/ui/test_errata.py [hunk 3]
```diff
@@ -131,11 +255,31 @@ def test_positive_apply_for_all_hosts(self):
...
+                    result = client.register_contenthost(
+                        self.session_org.label,
+                        self.activation_key.name,
+                    )
+                    self.assertEqual(result.return_code, 0)
+                    client.enable_repo(REPOS['rhst7']['id'])
+                    client.install_katello_agent()
+                    client.run(
+                        'yum install -y {0}'.format(FAKE_1_CUSTOM_PACKAGE))
+                with Session(self.browser):
```
Comment (renzon) [inline tests/foreman/ui/test_errata.py]: Too many nested context managers, maybe you could put [all of them together](http://stackoverflow.com/questions/3024925/python-create-a-with-block-on-several-context-managers).

File: tests/foreman/ui/test_errata.py [hunk 4]
```diff
@@ -244,15 +417,76 @@ def test_positive_filter(self):
...
         @Assert: Content hosts can be filtered by Environment.
 
-        @caseautomation: notautomated
-
-
-        @CaseLevel: Integration
+        @CaseLevel: System
         """
+        with VirtualMachine(distro='rhel71') as client1:
+            with VirtualMachine(distro='rhel71') as client2:
```
Comment (renzon) [inline tests/foreman/ui/test_errata.py]: Same about context managers nesting

File: tests/foreman/ui/test_errata.py [hunk 5]
```diff
@@ -348,15 +620,30 @@ def test_positive_content_host_library(self):
...
         @Assert: The errata from Library are displayed.
 
-        @caseautomation: notautomated
-
-
-        @CaseLevel: Integration
+        @CaseLevel: System
         """
+        with VirtualMachine(distro='rhel71') as client:
+            client.install_katello_ca()
```
Comment (renzon) [inline tests/foreman/ui/test_errata.py]: Same about context managers nesting

File: tests/foreman/ui/test_errata.py [hunk 6]
```diff
@@ -375,19 +662,95 @@ def test_positive_show_count_on_content_host_page(self):
...
+        })
+        ContentView.publish({'id': content_view.id})
+        cvv = ContentView.info({'id': content_view.id})['versions'][-1]
+        ContentView.version_promote({
+            'id': cvv['id'],
+            'organization-id': org.id,
+            'to-lifecycle-environment-id': env.id,
+        })
+        with VirtualMachine(distro='rhel67') as client:
+            client.install_katello_ca()
```
Comment (renzon) [inline tests/foreman/ui/test_errata.py]: Same about context managers nesting

File: tests/foreman/ui/test_errata.py [hunk 7]
```diff
@@ -408,11 +771,91 @@ def test_positive_show_count_on_content_host_details_page(self):
...
```
Comment (renzon) [inline tests/foreman/ui/test_errata.py]: Same about context managers nesting

**Comment by renzon** [comment]
ACK, @abalakh see only the context managers initialization once "Flat is better than nested". What a big PR, all the time I see this I think the same as @rochacbruno: Why not accept PR with more commits to keep history sane?

File: tests/foreman/ui/test_errata.py (hunk 3 above)
Comment (abalakh) [inline tests/foreman/ui/test_errata.py]: Thanks for suggestion, i'll apply it. But not in this case, for example - 2 VirtualMachines above can be combined, but Session should be started only when all the actions with those machines are finished successfully.

**Comment by abalakh** [comment]
Removed one problematic test from PR, will fix it and submit when i'm back from PTO. PR is ready to be merged

**Comment by oshtaier** [comment]
ACK

**Comment by rplevka** [comment]
this got to be cherry-picked to `6.2.z`

**Pull request merged**